pkg/jx/cmd/get_applications.go

```diff
@@ -243,6 +261,31 @@ func (o *GetApplicationsOptions) generateTable(apps []string, envApps []EnvApps,
 	return table
 }
 
+// BuildGitURL concatenates various fields of a SourceRepository CRD into a valid URL.
+func BuildGitURL(sourceRepository *v1.SourceRepository) string {
+	if sourceRepository.Spec.Provider == "" || sourceRepository.Spec.Org == "" || sourceRepository.Spec.Repo == "" {
+		return "None Found"
+	}
+	return sourceRepository.Spec.Provider + "/" + sourceRepository.Spec.Org + "/" + sourceRepository.Spec.Repo + ".git"
+}
+
+// GetSourceRepositoryForApplication fetches a SourceRepository CRD for a particular application by trying to match its name to a repo.
+func GetSourceRepositoryForApplication(application string, sourceRepositories altv1.SourceRepositoryInterface) (*v1.SourceRepository, error) {
+	// We can't pull the SourceRepository by name because we don't know the org, so we iterate over all of them and match the application name to the repo.
+	repoList, err := sourceRepositories.List(metav1.ListOptions{})
```

Comment:
I am wondering how efficient is this for a large number of repositories (e.g. 50+).

Reply:
That's a good question that I don't know the answer to, but I'm not sure there's another way to do this. You have to know the org + repo in order to build the SourceRepository name if you want to use Get for an individual SourceRepository. There's no way to get the org from the application, unless we were to change the naming of applications so that it's org + repo instead of just repo. Another solution is one unified CRD for apps and applications that references a SourceRepository. In my opinion this is the most robust solution, but obviously quite a bit more effort.

Reply:
+1 to unify the CRD for apps 

cc @pmuir

Reply:
Unifying apps and applications should be done but is a big job and completely orthogonal to this change - you still have to find the right crd for the name.

Anyway the solution to Cosmin's point is to use a label selector on the sourcerepository for already present `repository` label and then check the returned list which should be quite short.

You really need to get all the possible source repositories and handle the case that >1 exists.

Reply:
@pmuir That may reduce the size of the list returned in some cases, but I'm still not seeing a way to solve the ambiguity without reference to `owner` and `provider` as well, which is why I brought up putting them in a CRD, though it's a point well taken that unifying apps and applications is beside the point here. But adding a separate CRD would only add more work to that effort.

Reply:
> pmuir That may reduce the size of the list returned in some cases,

It will reduce it in every case except the one where every source repository you have has the same name but a different organisation. In the real world, that is every case ;-)

>  but I'm still not seeing a way to solve the ambiguity without reference to `owner` and `provider` as well, which is why I brought up putting them in a CRD, though it's a point well taken that unifying apps and applications is beside the point here. But adding a separate CRD would only add more work to that effort.

Yes. I see the point - I was confused as your solution suffers from the ambiguity ;-)

Probably a better way to handle this is via releases - look up the last release for the application (which you could do with a label and then sorting by version) and grab the git url

Reply:
@pmuir Okay, doing it by release makes sense, since the application name should match. I guess the same ambiguity is possible, for instance if you have a cluster with a bunch of forks of the same repo and each is a separate application. I'm not sure that's even possible, but it sounds like something that might happen in future, though to your point, this probably isn't a case to lose sleep over.

Reply:
The problem here is that releases, as far as I can tell, don't get labels, so a label selector is out. The application name is stored in the CRD but it's different from `Release.metadata.name`, so we'd need to do a field selector, except that `Release.spec.name` is not a supported field. So unless we can assume a version number (like 0.0.1 or something), which would allow us to select on `Release.metadata.name`, I'm officially stumped.